state/sqlserver/migration.go

```diff
@@ -273,92 +299,88 @@ func (m *migration) createStoredProcedureIfNotExists(db *sql.DB, name string, es
+				@RowVersion	BINARY(8),
+				@FirstWrite	BIT
+			) AS
+			BEGIN
 				IF (@FirstWrite=1)
 					BEGIN
 						IF (@RowVersion IS NOT NULL)
 							BEGIN
 								BEGIN TRANSACTION;
-								IF NOT EXISTS (SELECT * FROM [%s] WHERE [KEY]=@KEY AND RowVersion = @RowVersion)
+								IF NOT EXISTS (SELECT * FROM [%[3]s] WHERE [KEY]=@KEY AND RowVersion = @RowVersion)
 									BEGIN
 										THROW 2601, ''FIRST-WRITE: COMPETING RECORD ALREADY WRITTEN.'', 1
 									END
 								BEGIN
-									UPDATE [%s]
-									SET [Data]=@Data, UpdateDate=GETDATE()
+									UPDATE [%[3]s]
+									SET [Data]=@Data, UpdateDate=CURRENT_TIMESTAMP, ExpireDate=CASE WHEN @TTL IS NULL THEN NULL ELSE DATEADD(SECOND, @TTL, CURRENT_TIMESTAMP) END
 									WHERE [Key]=@Key AND RowVersion = @RowVersion
 								END
 								COMMIT;
 							END
 						ELSE
 							BEGIN
 								BEGIN TRANSACTION;
-								IF EXISTS (SELECT * FROM [%s] WHERE [KEY]=@KEY)
+								IF EXISTS (SELECT * FROM [%[3]s] WHERE [KEY]=@KEY)
 									BEGIN
 										THROW 2601, ''FIRST-WRITE: COMPETING RECORD ALREADY WRITTEN.'', 1
 									END
 								BEGIN
 									BEGIN TRY
-										INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
+										INSERT INTO [%[3]s] ([Key], [Data], ExpireDate) VALUES (@Key, @Data, CASE WHEN @TTL IS NULL THEN NULL ELSE DATEADD(SECOND, @TTL, CURRENT_TIMESTAMP) END)
 									END TRY
-						
+
 									BEGIN CATCH
 										IF ERROR_NUMBER() IN (2601, 2627)
-											UPDATE [%s]
-											SET [Data]=@Data, UpdateDate=GETDATE()
+											UPDATE [%[3]s]
+											SET [Data]=@Data, UpdateDate=CURRENT_TIMESTAMP, ExpireDate=CASE WHEN @TTL IS NULL THEN NULL ELSE DATEADD(SECOND, @TTL, CURRENT_TIMESTAMP) END
 											WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
 									END CATCH
 								END
-								COMMIT;	
+								COMMIT;
 							END
 					END
 				ELSE
 					BEGIN
 						IF (@RowVersion IS NOT NULL)
 							BEGIN
-								UPDATE [%s]
-								SET [Data]=@Data, UpdateDate=GETDATE()
+								UPDATE [%[3]s]
+								SET [Data]=@Data, UpdateDate=CURRENT_TIMESTAMP, ExpireDate=CASE WHEN @TTL IS NULL THEN NULL ELSE DATEADD(SECOND, @TTL, CURRENT_TIMESTAMP) END
 								WHERE [Key]=@Key AND RowVersion = @RowVersion
 								RETURN
 							END
 						ELSE
 							BEGIN
 								BEGIN TRY
-									INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
+									INSERT INTO [%[3]s] ([Key], [Data], ExpireDate) VALUES (@Key, @Data, CASE WHEN @TTL IS NULL THEN NULL ELSE DATEADD(SECOND, @TTL, CURRENT_TIMESTAMP) END)
 								END TRY
-					
+
 								BEGIN CATCH
 									IF ERROR_NUMBER() IN (2601, 2627)
-										UPDATE [%s]
-										SET [Data]=@Data, UpdateDate=GETDATE()
+										UPDATE [%[3]s]
+										SET [Data]=@Data, UpdateDate=CURRENT_TIMESTAMP, ExpireDate=CASE WHEN @TTL IS NULL THEN NULL ELSE DATEADD(SECOND, @TTL, CURRENT_TIMESTAMP) END
 										WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
```

Comment:
Probably needs a check here where if the row is expired (TTL is not NULL and is < current time), then it's not updated. Otherwise, this would return true for rows that are expired but not yet GC'd.

Reply:
Is this not the behavior we want? I understood we always SET the value, regardless of what the row currently is?

Reply:
I don't think so. Here's my thinking:

- A row that is expired is to be considered "deleted", even though it may not (yet) have been GC'd.
- This query uses UPDATE which means that it should fail if the row doesn't exist
- If we try to perform an update of a row that is expired, it should then fail (i.e. not update anything).

Reply:
You have not misunderstood, but this is a unique case.

If you are passing an Etag, you are telling Dapr to [observe the "first-write-wins" pattern](https://docs.dapr.io/developing-applications/building-blocks/state-management/state-management-overview/#why-dapr-chooses-optimistic-concurrency-control-occ) and you are instructing it to update a record only if the Etag matches.

In case of an expired row, although it's still in the database, it's "soft deleted", so the Etag should never match, and the user should get an error indicating Etag mismatch. With this query, a "soft deleted" record would be updated, however.

(I think the Postgres and SQLite implementation have the same bug, I can look into fixing those)

Reply:
Thanks for the explanation @ItalyPaleAle, I understand now!

Reply:
Thanks for the change.

I've updated Postgres, Cockroach, SQLite, and Oracle to fix this issue too: https://github.com/dapr/components-contrib/pull/2682#discussion_r1142299949